### PR TITLE
CPP: dbscheme annotations

### PR DESCRIPTION
**Pull request by geoffw0**
I have a very old note that we really should be using `keyset` more in the CPP dbscheme, as it's use should lead to better optimization (see https://semmle.com/wiki/display/IN/Schemas and page 8 of https://docs.google.com/presentation/d/1NhfjCmFIXin2Wi6QsU2Lw1Kc1ppUxJiHMYNZ1Wnx-30/edit?pli=1#slide=id.g293b9765be_0_158).  Since performance is on our minds at the moment (though possibly not really until Q2) I thought it would be a good time to raise the topic.

I've annotated about 10% of the dbscheme here as conversation starter.  This is not remotely ready to merge because:
 - I haven't created the required upgrade script
 - I haven't tested for correctness (`odasa checkDatabase`...)
 - I haven't tested for performance gains (or losses)

Do these changes look correct?  Useful?  What are your thoughts?

@ian-semmle @nickrolfe

### CODE REVIEW

**Comment by jbj** [comment]
These changes look alright to me, but they're of course in an area that doesn't matter for performance. They'd still be helpful as code documentation.

As for how we approach adding `keyset` annotations, I think most of the important db relations are minefields of potential inconsistencies that can arise in edge cases like when elements have the same name and/or location. This is closely tied to the extractor implementation. As long as our main motivation is speed, I can't justify aiming for completeness of `keyset` annotations. If we can get some key predicates like `getTarget()` and `getParent()` to be functional, that could give us a measurable difference in performance.

@ian-semmle, do you think that functional dependencies on the dbscheme will propagate through all the `unresolveElement` calls? Otherwise it seems pointless to add them to the dbscheme. What about our future plans to implement linker awareness in QL?

**Comment by ian-semmle** [comment]
I think they'll still be useful, but you'd have to ask an optimiser expert to be sure. At the very least they shouldn't hurt, provided they are consistent.

File: cpp/ql/src/semmlecode.cpp.dbscheme [hunk 1]
```diff
@@ -100,8 +103,9 @@ compilation_time(
  * message generated while extracting the `file_number`th file of that
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/semmlecode.cpp.dbscheme]: Well spotted, fixed.

**Comment by jbj** [comment]
@geoffw0 Could you also add a revert of #861? Cc @felicity-semmle and @jf205.

**Comment by geoffw0** [comment]
Rebased onto latest master.  Added an upgrade script in https://git.semmle.com/Semmle/code/pull/30828.

**Comment by geoffw0** [comment]
Just did a bit of testing locally on a small Windows project.  I can confirm that @dave-bartolomeo 's changes to `aggregate_field_init` and `aggregate_array_init` both resolve issues, other than that the results of `checkDatabase` are identical on builds with and without the changes in this PR.

Sadly, even on master, I encountered quite a number of other issues: `valuebind`, `is_proxy_class_for`, `decltypes` (need args 0 and 1), `derivedtypes` (need args 0 and 1), `var_decls` (need args 0 and 2), `fun_decls` (need args 0 and 2), and a bunch of `VALUE_NOT_IN_TYPE` errors I won't go into.  These are not caused by this PR.  @ian-semmle do we have any automated checks for this stuff?  On what platforms?

**Comment by ian-semmle** [comment]
Yes, qltest checks database consistency for all the testsuite databases on all platforms. We know that there are still problems in real-world databases; it just takes time to boil them down into a testcase and fix them.

**Comment by geoffw0** [comment]
OK, great (let me know if there's anything I can do to help with that).

**Comment by geoffw0** [comment]
I just built and `checkDatabase`d 'cmake' (via MinGW), with and without these changes.  Some problems with `aggregate_array_init` are fixed.  Some `[WARNING: INVALID_KEY]` problems with `valuebind`, `decltypes`, `enumconstants`, `globalvariables`, `membervariables`, `var_decls`, `fun_decls` exist before and after, and there are some `[VALUE_NOT_IN_TYPE]` errors.  There are no new problems.

I've not seen any problems.  Are we waiting for anything?  Should I retarget this to `rc/1.20`?

**Comment by geoffw0** [comment]
I've just removed the annotation on `externalData` because it caused a test failure.

**Pull request merged**